src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/IntellisenseQuickInfoBuilder.cs

```diff
@@ -47,18 +55,90 @@ internal static IntellisenseQuickInfoItem BuildItem(ITrackingSpan trackingSpan,
                 quickInfoItem.Sections.Where(s => s.Kind != QuickInfoSectionKinds.Description)
                                       .Select(BuildClassifiedTextElement));
 
+            // build text for RelatedSpan
+            if (quickInfoItem.RelatedSpans.Any())
+            {
+                elements.Add(await BuildTextElementForRelatedSpans(quickInfoItem, snapshot, document, cancellationToken).ConfigureAwait(false));
+            }
+
             var content = new ContainerElement(
                                 ContainerElementStyle.Stacked,
                                 elements);
 
             return new IntellisenseQuickInfoItem(trackingSpan, content);
         }
 
+        private static async Task<ClassifiedTextElement> BuildTextElementForRelatedSpans(CodeAnalysisQuickInfoItem quickInfoItem, ITextSnapshot snapshot, Document document, CancellationToken cancellationToken)
+        {
+            var classificationService = document.Project.LanguageServices.GetService<IClassificationService>();
+
+            var classifiedSpans = new List<ClassifiedSpan>();
+            var semanticSpans = new List<ClassifiedSpan>();
+
+            foreach (var span in quickInfoItem.RelatedSpans)
+            {
+                await classificationService.AddSyntacticClassificationsAsync(document, span, classifiedSpans, cancellationToken).ConfigureAwait(false);
+                await classificationService.AddSemanticClassificationsAsync(document, span, semanticSpans, cancellationToken).ConfigureAwait(false);
+            }
+
+            // replace the spans from SyntacticClassifications with spans from SemanticClassifications
+            classifiedSpans.Sort((a, b) => a.TextSpan.Start.CompareTo(b.TextSpan.Start));
+            semanticSpans.Sort((a, b) => a.TextSpan.Start.CompareTo(b.TextSpan.Start));
+
+            var i = 0;
+            foreach (var sSpan in semanticSpans)
+            {
+                for (; i < classifiedSpans.Count; i++)
```

Comment:
Is the assumption here that any semanticspan will always have at least one match in classifiedSpans? My worry is if there's one that doesn't, then `i` will advance to the end of the list and skip other matches...